nautobot_circuit_maintenance/handle_notifications/sources.py

```diff
@@ -247,58 +327,168 @@ def receive_notifications(self, logger: Job, since: int = None) -> Iterable[Main
+            self.credentials.refresh(Request())
+
+    def build_service(self):
+        """Build API service."""
+        self.service = build("gmail", "v1", credentials=self.credentials)
+
+    def close_service(self):
+        """Close API Service."""
+        if self.service:
+            self.service.close()
+
+    def _authentication_logic(self):
+        """Inner method to run the custom class validation logic."""
+        self.load_credentials()
+
+    def extract_raw_payload(self, body: Dict, msg_id: bytes) -> bytes:
+        """Extracts the raw_payload from body or attachement."""
+        if "attachmentId" in body:
+            attachment = (
+                self.service.users()  # pylint: disable=no-member
+                .messages()
+                .attachments()
+                .get(userId=self.account, messageId=msg_id, id=body["attachmentId"])
+                .execute()
             )
-            return False
+            return base64.b64decode(attachment["data"])
+        if "data" in body:
+            return base64.urlsafe_b64decode(body["data"])
 
-        for provider in notification_source.providers.all():
-            cm_cf = CustomField.objects.get(name="emails_circuit_maintenances")
-            provider_emails = provider.get_custom_fields().get(cm_cf)
-            if provider_emails:
-                self.emails_to_fetch.extend(provider_emails.split(","))
-                providers_with_email.append(provider.name)
-            else:
-                providers_without_email.append(provider.name)
+        return b""
 
-        if providers_without_email:
+    # pylint: disable=inconsistent-return-statements,too-many-locals,too-many-branches
+    def fetch_email(self, logger: Job, msg_id: bytes, since: Optional[int]) -> Optional[MaintenanceNotification]:
+        """Fetch an specific email ID.
+
+        See data format:  https://developers.google.com/gmail/api/reference/rest/v1/users.messages#Message
+        """
+        received_email = (
+            self.service.users().messages().get(userId=self.account, id=msg_id).execute()  # pylint: disable=no-member
+        )
+        email_subject = ""
+        email_source = ""
+
+        for header in received_email["payload"]["headers"]:
+            if header.get("name") == "Subject":
+                email_subject = header["value"]
+            elif header.get("name") == "From":
+                email_source = header["value"]
+
+        if not email_subject or not email_source:
             logger.log_warning(
-                message=(
-                    f"Skipping {', '.join(providers_without_email)} because these providers have no email configured."
-                ),
+                received_email, message="Received email doesn't contain either Suject or From, so skipping it."
             )
+            return
 
-        if not providers_with_email:
+        if since:
+            if received_email["internalDate"] < since:
```

Comment:
Just confirming - `received_email["internalData"]` is already a properly defined `int` suitable for comparison with `since`, no conversion required?

Reply:
good spot, it wasn't https://developers.google.com/gmail/api/reference/rest/v1/users.messages#Message